### PR TITLE
feat: HTTPS local registry

### PR DESCRIPTION
**Pull request by Skarlso**
# Enabling HTTPS for the in-cluster registry

The in-cluster registry is required to run in HTTPS mode. This requires a few setups, secrets and generated certificates. In the development environment, this is already taken care of.

Secrets can be defined for production environments so the controllers can talk to the HTTPS registry.

## Configuration

A set of Certificates needs to be provided to configure the HTTPS registry. There are three locations where certificates need configuration.

First, the Registry. The volume patch is located under `config/registry/patches`. 

```yaml
      containers:
        - name: registry
          image: registry:2
          env:
            - name: REGISTRY_HTTP_TLS_CERTIFICATE
              value: "/certs/cert.pem"
            - name: REGISTRY_HTTP_TLS_KEY
              value: "/certs/key.pem"
            - name: REGISTRY_HTTP_TLS_CLIENTCAS_0
              value: "/certs/ca.pem"
          volumeMounts:
            - mountPath: "/certs"
              name: "registry-certs"
      volumes:
        - name: "registry-certs"
          secret:
            secretName: "registry-cert"
            items:
              - key: "cert.pem"
                path: "cert.pem"
              - key: "key.pem"
                path: "key.pem"
              - key: "ca.pem"
                path: "ca.pem"
```

Second, the Controller.

The developer certificates share the server and the client certs. This is a convenience. In product environments, you can provide the server certs to the registry and the client certs to the controllers.

The controller also has a patch like the above under `config/manager/patches`:

```yaml
      containers:
        - name: manager
          image: open-component-model/ocm-controller
          volumeMounts:
            - mountPath: "/certs"
              name: "certificates"
      volumes:
        - name: "certificates"
          secret:
            secretName: "registry-certs"
            items:
              - key: "caFile"
                path: "ca.pem"
```

Another optional setting on the controller is the name of the secret that contains the certificates:

```go
flag.StringVar(&ociRegistryCertSecretName, "certificate-secret-name", v1alpha1.DefaultRegistryCertificateSecretName, "")
```

The default location is `/certs`.

## Generating Certs / Using Trusted Certificates

There are many ways of generating a certificate. With openssl, [cfssl](https://github.com/cloudflare/cfssl), [mkcert](https://github.com/FiloSottile/mkcert) and others.

The ocm-controller chooses to build developer certificates using `mkcert`. However the certificates are provided, they must be trusted by the system that uses the controller.

We are doing this in the controllers and on CI via an `entrypoint.sh` script. This script will take the root certificate from a specific location ( later on could be configurable via environment properties passed to the controller ) and append it to the `ca-certificates.crt` container root certificate. This is to ensure that the controller trusts the certificate provided to it.

## Creating Secrets

Once the certificates are generated, create and apply the secret containing them. We using `kubectl` to do this.

```bash
kubectl create secret generic \
  -n ocm-system registry-certs \
  --from-file=ca.pem="certs/ca.pem" \
  --from-file=cert.pem="certs/cert.pem" \
  --from-file=key.pem="certs/key.pem" \
  --dry-run=client -o yaml > ./registry_certs_secret.yaml
```

## Developer Certificates

We need two things to run tests like our e2e tests on CI or locally. First, the generated self-signed certificate must be trusted locally, and second, the GitHub action runner environment.

## Tilt

Tilt will handle everything for the local user once the secret exists. For local development, when not running `e2e` tests with `make e2e` the secret has to be generated before `tilt up` can be executed.

To generate the secret simply run `./hack/create_developer_certificate_secrets.sh`. This will do two things. First, generate some certificates under `hack/certs` and second, generate the secret Tilt will use to apply to the cluster. Then, `tilt up` can be run as before.

When running the `make e2e` target, it will ensure that `mkcert` is installed, `mkcert -install` is executed, and that the secrets exist and are applied to the cluster. No additional actions will have to be done by the user when running end-to-end tests.

## CI

The observant eye will see that the script for generating certificates under hack is full of `sudo` paragraphs. This is because GitHub action containers will not update the certificate unless run with sudo, and `mkcert` will fail to install the certs at the system trust store.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -7,6 +7,8 @@ replace google.golang.org/grpc => google.golang.org/grpc v1.29.0
 
 replace github.com/opencontainers/go-digest => github.com/opencontainers/go-digest v1.0.1-0.20220411205349-bde1400a84be
 
+replace github.com/open-component-model/ocm-e2e-framework => ../ocm-e2e-framework
```
Comment (Skarlso) [inline go.mod:10]: Remove this!

Added this because adding a component now requires https. I will have to fix that in the framework.

**Comment by Skarlso** [comment]
Finally works locally. There are some adjustments left on unit tests and the e2e test.

Screenshot 2023-06-27 at 15 31 29

File: controllers/componentversion_controller.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ type ComponentVersionReconciler struct {
 //+kubebuilder:rbac:groups="",resources=services;pods,verbs=get;create;update;patch;delete
 //+kubebuilder:rbac:groups="apps",resources=deployments,verbs=get;create;update;patch;delete
 
-// +kubebuilder:rbac:groups="",resources=secrets;serviceaccounts,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;serviceaccounts,verbs=create;get;list;watch
```
Comment (phoban01) [inline controllers/componentversion_controller.go:55]: the component version controller shouldn't need to create a service account, should it?

Comment (phoban01) [reply controllers/componentversion_controller.go:55]: I think we can remove the create here as it doesn't need to create secrets or service accounts.

File: hack/create_developer_certificate_secrets.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+fi
+
+echo "checking if secrets have been correctly generated"
+
+# Set up certificate paths
+serverPemPath="./hack/certs/server.pem"
+serverKeyPath="./hack/certs/server-key.pem"
+rootCAPath=$(./bin/mkcert -CAROOT)/rootCA.pem
+
+if [ ! -e "${serverPemPath}" ] && [ ! -e "${serverKeyPath}" ]; then
```
Comment (phoban01) [inline hack/create_developer_certificate_secrets.sh:24]: could we just generate if not found?

File: config/manager/patches/add_root_certificates.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+              - key: "ca-certificates.crt"
+                path: "ca-certificates.crt"
+        - name: "client-certificates"
+          secret:
+            secretName: "registry-certs"
+            items:
+              - key: "server.pem"
+                path: "server.pem"
+              - key: "server-key.pem"
+                path: "server-key.pem"
```
Comment (phoban01) [inline config/manager/patches/add_root_certificates.yaml:35]: Any reason to not put these all in the same secret? That would be how I usually set TLS secrets configured.

Comment (Skarlso) [reply config/manager/patches/add_root_certificates.yaml:35]: Because, strictly, the root certificate is not needed in the registry container.

File: config/manager/kustomization.yaml [hunk 5]
```diff
@@ -2,3 +2,7 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 resources:
 - deployment.yaml
+
+# Uncomment to enable HTTPS for the registry
+#patches:
+#- path: ./patches/add_root_certificates.yaml
```
Comment (souleb) [inline config/manager/kustomization.yaml:8]: is it working without the patch? You already try to mount the secret in `deployment.yaml` above.

File: config/manager/patches/add_root_certificates.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+          secret:
+            secretName: "developer-root-certificate"
+            items:
+              - key: "ca-certificates.crt"
+                path: "ca-certificates.crt"
+        - name: "client-certificates"
+          secret:
+            secretName: "registry-certs"
+            items:
+              - key: "server.pem"
```
Comment (souleb) [inline config/manager/patches/add_root_certificates.yaml:32]: for e2e it's fine to provide same `pem` for server and client, but for end users I think those should be different.

File: controllers/componentversion_controller.go (hunk 2 above)
Comment (Skarlso) [reply controllers/componentversion_controller.go:55]: Yeah. I added create for the `secret` a commit ago when we were creating the cert secret. But now, we just expect it to be there. So yeah, I can remove the `create` here.

File: hack/create_developer_certificate_secrets.sh (hunk 3 above)
Comment (Skarlso) [reply hack/create_developer_certificate_secrets.sh:24]: Actually, I can combine the two and make it one target. Hah.

File: config/registry/patches/add_registry_certificates.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+              value: "/certs/server.pem"
+            - name: REGISTRY_HTTP_TLS_KEY
+              value: "/certs/server-key.pem"
+          volumeMounts:
+            - mountPath: "/certs"
+              name: "registry-certs"
+      volumes:
+        - name: "registry-certs"
+          secret:
+            secretName: "registry-cert"
```
Comment (souleb) [inline config/registry/patches/add_registry_certificates.yaml:31]: I think the registry needs the `CA` too in order to validate client certs.

File: main.go [hunk 8]
```diff
@@ -48,16 +49,24 @@ func init() {
...
+		ociRegistryCertificateLocation string
+		ociRegistryInsecureSkipVerify  bool
+		ociRegistryNamespace           string
+	)
 
 	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
 	flag.StringVar(&eventsAddr, "events-addr", "", "The address of the events receiver.")
 	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
 	flag.StringVar(&ociRegistryAddr, "oci-registry-addr", ":5000", "The address of the OCI registry.")
+	flag.StringVar(&ociRegistryCertificateLocation, "oci-registry-certificate-location", v1alpha1.DefaultRegistryCertificateLocation, "The location in the contain where certificates live for the registry.")
```
Comment (souleb) [inline main.go:67]: can we use those instead:
```
       --ca-file string             certificate authority file path
       --cert-file string           certificate file path
       --key-file string            key file path
```

People are use to seeing those, and it's easier to understand what's requested.

File: pkg/oci/repository.go [hunk 9]
```diff
@@ -51,16 +56,105 @@ type ResourceOptions struct {
...
 // Client implements the caching layer and the OCI layer.
 type Client struct {
-	OCIRepositoryAddr string
+	Client              client.Client
+	OCIRepositoryAddr   string
+	CertificateLocation string
+	InsecureSkipVerify  bool
+	Namespace           string
+
+	serverPem    []byte
```
Comment (souleb) [inline pkg/oci/repository.go:99]:
```suggestion
	cert    []byte
```

File: pkg/oci/repository.go [hunk 10]
```diff
@@ -51,16 +56,105 @@ type ResourceOptions struct {
...
+	serverKeyPem []byte
```
Comment (souleb) [inline pkg/oci/repository.go:100]:
```suggestion
	key []byte
```

Those are used in client side right?

File: pkg/oci/repository.go [hunk 11]
```diff
@@ -51,16 +56,105 @@ type ResourceOptions struct {
...
+			serverKeyPem, err := os.ReadFile(filepath.Join(c.CertificateLocation, "server-key.pem"))
+			if err != nil {
+				return fmt.Errorf("server-key.pem data not found in registry certificate secret: %w", err)
+			}
+
+			c.serverPem = serverPem
+			c.serverKeyPem = serverKeyPem
+		}
+
+		tlsConfig.Certificates = []tls.Certificate{
```
Comment (souleb) [inline pkg/oci/repository.go:127]: what about `CA`?

File: config/manager/patches/add_root_certificates.yaml (hunk 6 above)
Comment (Skarlso) [reply config/manager/patches/add_root_certificates.yaml:32]: That is none of our concern though, right?

File: pkg/oci/repository.go (hunk 11 above)
Comment (Skarlso) [reply pkg/oci/repository.go:127]: I will add that.

File: config/manager/kustomization.yaml (hunk 5 above)
Comment (Skarlso) [reply config/manager/kustomization.yaml:8]: I forgot to remove it. :D

Comment (Skarlso) [reply config/manager/kustomization.yaml:8]: But it does work, yeah.

Comment (Skarlso) [reply config/manager/kustomization.yaml:8]: Verified by running: `kustomize build ./config/default > ./install.yaml`.

File: config/manager/patches/add_root_certificates.yaml (hunk 4 above)
Comment (Skarlso) [reply config/manager/patches/add_root_certificates.yaml:35]: I guess I can add it in one secret.

File: config/manager/patches/add_root_certificates.yaml (hunk 6 above)
Comment (souleb) [reply config/manager/patches/add_root_certificates.yaml]: I guess not here, but on cli side yeah.

File: controllers/componentversion_controller.go (hunk 2 above)
Comment (phoban01) [reply controllers/componentversion_controller.go:55]:
```suggestion
// +kubebuilder:rbac:groups="",resources=secrets;serviceaccounts,verbs=get;list;watch
```

**Review by phoban01 — COMMENTED** [review]
Overall looks good.

File: controllers/componentversion_controller.go (hunk 2 above)
Comment (Skarlso) [reply controllers/componentversion_controller.go:55]: Yah, I still have to do this. :D I won't forget. Doing it now.

File: pkg/ocm/ocm.go [hunk 12]
```diff
@@ -418,3 +397,86 @@ func HashIdentity(id ocmmetav1.Identity) (string, error) {
...
+	defer ociRepo.Close()
+
+	genericRepo, err := genericocireg.NewRepository(octx, genericocireg.DefaultComponentRepositoryMeta(nil), ociRepo)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get repository for spec: %w", err)
+	}
+	defer genericRepo.Close()
+
+	// get the component Version
+	// has to be closed at calling site.
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply pkg/ocm/ocm.go:464]: Good idea!

File: e2e/suite_test.go [hunk 13]
```diff
@@ -59,3 +64,30 @@ func TestMain(m *testing.M) {
 
 	os.Exit(testEnv.Run(m))
 }
+
+func dumpControllerLogs(n string) env.Func {
```
Comment (Skarlso) [inline e2e/suite_test.go:68]: @phoban01 @souleb  I added this as a debug. Do you guys think I should make this a thing in ocm-e2e-controller instead? Is it worth it, or should I just delete it?

File: pkg/oci/repository.go [hunk 14]
```diff
@@ -51,16 +47,134 @@ type ResourceOptions struct {
...
+func WithClient(client client.Client) ClientOptsFunc {
+	return func(opts *Client) {
+		opts.Client = client
+	}
+}
+
 // Client implements the caching layer and the OCI layer.
 type Client struct {
-	OCIRepositoryAddr string
+	Client             client.Client
```
Comment (Skarlso) [inline pkg/oci/repository.go:98]: I just realised I don't need the client anymore.

File: pkg/ocm/ocm.go [hunk 15]
```diff
@@ -43,24 +46,39 @@ type Contract interface {
 	GetResource(ctx context.Context, octx ocm.Context, cv *v1alpha1.ComponentVersion, resource *v1alpha1.ResourceReference) (io.ReadCloser, string, error)
 	GetComponentVersion(ctx context.Context, octx ocm.Context, obj *v1alpha1.ComponentVersion, name, version string) (ocm.ComponentVersionAccess, error)
 	GetLatestValidComponentVersion(ctx context.Context, octx ocm.Context, obj *v1alpha1.ComponentVersion) (string, error)
-	ListComponentVersions(logger logr.Logger, octx ocm.Context, obj *v1alpha1.ComponentVersion) ([]Version, error)
```
Comment (Skarlso) [inline pkg/ocm/ocm.go:46]: There was really no need for this to be in the public API.

File: pkg/ocm/ocm.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Skarlso) [reply pkg/ocm/ocm.go:494]: Added a close function that can be called on the componentBundle instead so at least it's transparent.

**Comment by Skarlso** [comment]
@Skarlso Let's not forget that after this is done, I need to build and push a new version of Repository.

File: go.mod [hunk 16]
```diff
@@ -2,9 +2,6 @@ module github.com/open-component-model/ocm-controller
 
 go 1.20
 
-// Something requires the naming package. :( Which was an experimental package and has been removed.
-replace google.golang.org/grpc => google.golang.org/grpc v1.29.0
```
Comment (Skarlso) [inline go.mod:6]: Turns out, this is no longer needed. 🎉

File: e2e/suite_test.go (hunk 13 above)
Comment (souleb) [reply e2e/suite_test.go:68]: this would be valuable in the test framework as a wait to check status. But maybe not pods but and deployments? and Using a kubeClient instead of shelling out?

File: hack/create_developer_certificate_secrets.sh [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+# Set up certificate paths
+# Sudo is required for the install to work in GitHub actions.
+sudo ./bin/mkcert -install
+certPath="./hack/certs/cert.pem"
+keyPath="./hack/certs/key.pem"
+rootCAPath=$(sudo ./bin/mkcert -CAROOT)/rootCA.pem
+
+echo "updating root certificate"
+
+sudo cat "${rootCAPath}" | sudo tee -a /etc/ssl/certs/ca-certificates.crt || echo "appending to ca-certificates failed but ignoring"
```
Comment (souleb) [inline hack/create_developer_certificate_secrets.sh:28]:
```suggestion
sudo cat "${rootCAPath}" | sudo tee -a /etc/ssl/certs/ca-certificates.crt || echo "failed to append to ca-certificates. Ignoring the failure"
```

File: hack/entrypoint.sh [hunk 18]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env sh
+
+rootCA=/certs/ca.pem
+
+if [ ! -e "${rootCA}" ]; then
+  echo "root certificate at location ${rootCA} not found, ignoring appending it..."
```
Comment (souleb) [inline hack/entrypoint.sh:6]:
```suggestion
  echo "warning... root certificate at location ${rootCA} not found."
```

File: hack/entrypoint.sh [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+  exec "$@"
+fi
+
+echo "Updating root certificate with provided certificate..."
```
Comment (souleb) [inline hack/entrypoint.sh:11]:
```suggestion
echo "updating root certificate with provided certificate..."
```

File: pkg/ocm/ocm.go [hunk 20]
```diff
@@ -43,24 +46,39 @@ type Contract interface {
...
+	cache        cache.Cache
+	disableHttps bool
 }
 
 var _ Contract = &Client{}
 
+type ClientOptionsFunc func(c *Client)
+
+// WithDisableHTTPS disables the https repository setting.
+func WithDisableHTTPS() ClientOptionsFunc {
```
Comment (souleb) [inline pkg/ocm/ocm.go:64]:
```suggestion
func WithDisabledHTTPS() ClientOptionsFunc {
```

File: pkg/ocm/ocm.go [hunk 21]
```diff
@@ -43,24 +46,39 @@ type Contract interface {
...
 	VerifyComponent(ctx context.Context, octx ocm.Context, obj *v1alpha1.ComponentVersion, version string) (bool, error)
 }
 
 // Client implements the OCM fetcher interface.
 type Client struct {
-	client client.Client
-	cache  cache.Cache
+	client       client.Client
+	cache        cache.Cache
+	disableHttps bool
```
Comment (souleb) [inline pkg/ocm/ocm.go:56]:
```suggestion
	disabledHttps bool
```

File: pkg/ocm/ocm.go [hunk 22]
```diff
@@ -418,3 +414,99 @@ func HashIdentity(id ocmmetav1.Identity) (string, error) {
 
 	return fmt.Sprintf("sha-%d", hash), nil
 }
+
+type componentBundle struct {
+	versioned ocm.ComponentVersionAccess
+	list      ocm.ComponentAccess
+}
+
+func (c *componentBundle) Close() error {
```
Comment (souleb) [inline pkg/ocm/ocm.go:423]: should this be exported?

File: pkg/ocm/ocm.go [hunk 23]
```diff
@@ -418,3 +414,99 @@ func HashIdentity(id ocmmetav1.Identity) (string, error) {
...
+	if c.versioned != nil {
+		return c.versioned.Close()
+	}
+
+	return c.list.Close()
+}
+
+type lookupComponentOption struct {
```
Comment (souleb) [inline pkg/ocm/ocm.go:431]: if this is to be internal only maybe we don't need the functional option here. We can refactor those without breaking consumers code.

File: pkg/ocm/ocm.go (hunk 22 above)
Comment (Skarlso) [reply pkg/ocm/ocm.go:423]: ah yeah, no, it doesn't need to be

File: pkg/ocm/ocm.go (hunk 23 above)
Comment (Skarlso) [reply pkg/ocm/ocm.go:431]: The functional options are there because sometimes it's called with a version and sometimes not. Sometimes there is a name sometimes not. And I think passing in `""` just looked ugly. :D So that's why I choose these options. :) WDYT?

File: e2e/suite_test.go (hunk 13 above)
Comment (Skarlso) [reply e2e/suite_test.go:68]: Yeah, that's what I was thinking as well. I'll implement it there. :)

Comment (Skarlso) [reply e2e/suite_test.go:68]: Also, I don't mind using kubectl. It's a hell of a lot easier. This is a test framework. I see no point in complicating it. After all, the main Kuberentes e2e-framework is using kind's CLI to stop start kind clusters.

Comment (Skarlso) [reply e2e/suite_test.go:68]: But I'm open for change. Do you know any benefits about using the API instead of shelling out?

Comment (souleb) [reply e2e/suite_test.go:68]: the only advantage I see is that you don't need to install those binaries in your pipeline

File: pkg/ocm/ocm.go (hunk 23 above)
Comment (souleb) [reply pkg/ocm/ocm.go:431]: functional options are used to avoid breaking api contract. Here you clearly don't have that use case. You can just pass in the struct like:

```go
func (c *Client) lookupComponent(ctx context.Context, octx ocm.Context, obj *v1alpha1.ComponentVersion, opts lookupComponentOption) (*componentBundle, error) {
```

this way you don't need the extra `funcs`

File: e2e/suite_test.go (hunk 13 above)
Comment (Skarlso) [reply e2e/suite_test.go:68]: But you already have a Kubernetes cluster so 95% chance you already have kubectl. :D For us, it's a 100% since it's already required anyways. :D 

Anyways, I already wrote the code to stream logs so I can actually just copy it over I guess.

I'll use the library. :)

**Comment by phoban01** [comment]
Last request from my is that we add a decent overview to the PR description. I can use this as a basis to update the installation docs for the controller on the website with a section regarding configuring HTTPS.

**Comment by Skarlso** [comment]
@phoban01 Absolutely. I'm writing it up.

File: pkg/ocm/ocm.go (hunk 23 above)
Comment (Skarlso) [reply pkg/ocm/ocm.go:431]: Functional options are not just for breaking apis. They are also there for slick API design with variable settings and in the mind of "define it only when needed". With the struct, you always have to pass in the struct. Here for example:

```go
func (c *Client) listComponentVersions(logger logr.Logger, octx ocm.Context, obj *v1alpha1.ComponentVersion) ([]Version, error) {
	cv, err := c.lookupComponent(context.Background(), octx, obj)
	if err != nil {
		return nil, fmt.Errorf("failed to lookup componint: %w", err)
	}
```

It would mean I would have to pass in an empty struct without reason like `lookupComponentOption{}`. Which is ugly IMHO. Since the name doesn't need to be overwritten and we don't require a version.

**Comment by Skarlso** [comment]
@phoban01 Added a huge description. I hope it's okay?

File: pkg/ocm/ocm.go (hunk 23 above)
Comment (souleb) [reply pkg/ocm/ocm.go:431]: I understand your point of view. It would save you 2 `withFunc` functions though. Your choice.

Comment (Skarlso) [reply pkg/ocm/ocm.go:431]: I'll think about it. :D I'll see how I like it. :)

**Comment by Skarlso** [comment]
__Note__: adding certs to OciRepository.

**Comment by Skarlso** [comment]
I'll update the description once I implement providing a secret name instead.

File: pkg/ocm/ocm.go (hunk 23 above)
Comment (Skarlso) [reply pkg/ocm/ocm.go:431]: Alright, I re-wrote it. :) It does look better, and less code. :)

File: e2e/testdata/testOCMController/component_version.yaml [hunk 24]
```diff
@@ -7,8 +7,8 @@ spec:
   component: component.name.pipeline/podinfo
   interval: 1m
   references:
-    expand: true  
+    expand: true
   repository:
-    url: http://registry.ocm-system.svc.cluster.local:5000
+    url: registry.ocm-system.svc.cluster.local:5000
```
Comment (Skarlso) [inline e2e/testdata/testOCMController/component_version.yaml:12]: This is a necessary change. Otherwise, OCM parses the repo URL as a plain URL and doesn't set the scheme to the appropriate value.

File: api/v1alpha1/componentversion_types.go [hunk 25]
```diff
@@ -61,6 +61,7 @@ type ComponentVersionSpec struct {
 // Repository specifies access details for the repository that contains OCM ComponentVersions.
 type Repository struct {
 	// URL specifies the URL of the OCI registry in which the ComponentVersion is stored.
+	// MUST NOT CONTAIN THE SCHEME.
```
Comment (Skarlso) [inline api/v1alpha1/componentversion_types.go:64]: @phoban01 @souleb I'm thinking about enforcing this by a webhook regex rule.

**Comment by Skarlso** [comment]
MPAS flow is working.

Relevant PRs are tagged.

File: api/v1alpha1/componentversion_types.go (hunk 25 above)
Comment (souleb) [reply api/v1alpha1/componentversion_types.go:64]: yes if we plan to enforce https, better to fail early.

File: controllers/fluxdeployer_controller.go [hunk 26]
```diff
@@ -49,6 +49,8 @@ type FluxDeployerReconciler struct {
 	RegistryServiceName string
 	RetryInterval       time.Duration
 	DynamicClient       dynamic.Interface
+
+	CertSecretName string
```
Comment (souleb) [inline controllers/fluxdeployer_controller.go:53]: why do we need this here? If it is to permit passing arbitrary `certs` I should rather be another PR.

File: controllers/snapshot_controller.go [hunk 27]
```diff
@@ -131,7 +131,7 @@ func (r *SnapshotReconciler) Reconcile(ctx context.Context, req ctrl.Request) (r
 
 	obj.Status.LastReconciledDigest = obj.Spec.Digest
 	obj.Status.LastReconciledTag = obj.Spec.Tag
-	obj.Status.RepositoryURL = fmt.Sprintf("http://%s/%s", r.RegistryServiceName, name)
+	obj.Status.RepositoryURL = fmt.Sprintf("https://%s/%s", r.RegistryServiceName, name)
```
Comment (souleb) [inline controllers/snapshot_controller.go:134]: maybe set a `const scheme = "https"` so we can change it easily in future?

File: e2e/testdata/testOCMController/component_version.yaml (hunk 24 above)
Comment (souleb) [reply e2e/testdata/testOCMController/component_version.yaml:12]: but this is coming... we could leave a `todo` here.

File: pkg/oci/repository.go [hunk 28]
```diff
@@ -51,16 +48,119 @@ type ResourceOptions struct {
...
+	Namespace          string
+	CertSecretName     string
+
+	certPem []byte
+	keyPem  []byte
+	ca      []byte
+}
+
+// WithTransport sets up insecure TLS so the library is forced to use HTTPS.
+func (c *Client) WithTransport() Option {
```
Comment (souleb) [inline pkg/oci/repository.go:97]: we should split this function, it does too many things

**Review by souleb — APPROVED** [review]
LGTM! awesome job @Skarlso 🥇

File: pkg/oci/repository.go (hunk 28 above)
Comment (Skarlso) [reply pkg/oci/repository.go:97]: 👍

File: controllers/fluxdeployer_controller.go (hunk 26 above)
Comment (Skarlso) [reply controllers/fluxdeployer_controller.go:53]: Talked about this on Slack. Because the OCIRepository requires the certificate otherwise it would need an `insecure` setting.

**Comment by Skarlso** [comment]
> yes if we plan to enforce https, better to fail early.

This should not be needed once we apply latest OCM. That won't happen in this PR though to avoid breaking changes form the OCM library. We'll apply the fix in a follow-up.
